### ISSUE TITLE
use record count for hive table hash

### ISSUE THREAD

**Issue by AliTajeldin** [TresAmigosSD/SMV#1088]
Currently, if new records are added to a hive table, SMV will not recompute downstream modules as the hash of the Hive Table module will not change.  There is not direct way to get the "last update time" of a hive table.  Therefore, we will use the record count as a proxy for that.

Need to validate that computing the count is fast regardless of the data format (csv, ORC, parquet)

**Comment by jacobdr** [comment]
Good to make this a pluggable “strategy”. Partition diffing (or change in the number of partitions), max date of a column changing, and modification time of the files making up the table are all possible alternatives we could use (off the top of my head...)

**Comment by AliTajeldin** [comment]
Let's keep this simple for now if the count proves efficient in all backing store types.

**Comment by jacobdr** [comment]
Agreed... Whoever implements this be mindful of the need to potentially refresh Hive table stats to get the proper record count. For ORC/Parquet counts are stored in the files themselves, for CSV I think we need to go to the metastore

**Comment by laneb** [comment]
Have we explored whether we can get Hive to provide the paths to all of the files backing the table? If so, maybe we can just check the count/timestamps of the files.

**Comment by AliTajeldin** [comment]
I think there is a way to get that info through the `HCatalog`.  We would also need to handle case where we would not have access to underlying HDFS files (somewhat common setup, btw).

**Comment by ninjapapa** [comment]
I ran into a case in a client project,

* Most client tables updated monthly, while a small group updated weekly
* Each week the weekly table hive table get updated, and a column called `dt_ld_time` in the new table of all rows get re-written
* Client want to have weekly runs which only apply the need updates instead of full `--force-run-all`, which is what we told them for the monthly run
* Client don't want to change code for their weekly runs

There could be 2 approaches for this case:
* Have a way to let input hive module's hash depend on some user specified field (first record's value)
* Have a way for user to specify input table which he knew changed

**Comment by AliTajeldin** [comment]
@ninjapapa in the scenario above, when the weekly tables are updates, is the **entire** table updated or just some rows?  That is do rows in the table potentially have different `dt_ld_time` values?  And if they do, do we need to essentially filter for only the latest `dt_ld_time` and do the run on these records or just re-run against the entire table?

**Comment by ninjapapa** [comment]
* Entire table.
* The same for this client. Not sure whether it is by design or just by accident. 
* Re-run on the entire table is needed

**Comment by laneb** [comment]
Would this requirement not be accounted for if we were checking the underlying HDFS files?

**Comment by ninjapapa** [comment]
I think we should have a way for user to specify 'changed' input at run time with or without figure out how to calculate the hash based on file on HDFS. Since there could be permission issue to access the underlying files.

**Comment by AliTajeldin** [comment]
@laneb @ninjapapa Let's have a brainstorming session when we are all in office.
